src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java

```diff
@@ -486,12 +468,105 @@ public static void applyWorkspaceEdit(@NotNull WorkspaceEdit edit,
+        Range range = textEdit.getRange();
+        if (range == null) {
+            return;
+        }
+        String text = textEdit.getNewText();
+//        String newText = StringUtilRt.convertLineSeparators(textEdit.getNewText());
+        int start = toOffset(range.getStart(), document);
+        int end = toOffset(range.getEnd(), document);
+  //      document.replaceString(startOffset, endOffset, newText);
+
+        if (StringUtils.isEmpty(text)) {
+            document.deleteString(start, end);
+        } else {
+            text = text.replaceAll("\r", "");
+            if (end >= 0) {
+                if (end - start <= 0) {
+                    document.insertString(start, text);
+                } else {
+                    document.replaceString(start, end, text);
+                }
+            } else if (start == 0) {
+                document.setText(text);
+            } else if (start > 0) {
+                document.insertString(start, text);
+            }
+        }
+    }
+
+    private static void applyCreateFile(CreateFile createFile) {
+        VirtualFile targetFile = findResourceFor(createFile.getUri());
+        if (targetFile != null && createFile.getOptions() != null) {
+            if (!createFile.getOptions().getIgnoreIfExists()) {
+                Document document = getDocument(targetFile);
+                if (document != null) {
+                    TextEdit textEdit = new TextEdit(new Range(toPosition(0, document), toPosition(document.getTextLength(), document)), "");
+                    applyTextEdits(document, Collections.singletonList(textEdit));
+                }
+            }
+        } else {
+            try {
+                String fileUri = createFile.getUri();
+                createFile(fileUri);
+            } catch (IOException e) {
+                LOGGER.warn(e.getLocalizedMessage(), e);
+            }
+        }
+    }
 
+    private static void applyDeleteFile(DeleteFile deleteFile) {
+        try {
+            VirtualFile resource = findResourceFor(deleteFile.getUri());
+            if (resource != null) {
+                resource.delete(null);
             }
+        } catch (IOException e) {
+            LOGGER.warn(e.getLocalizedMessage(), e);
+        }
+    }
 
+    private static void applyRenameFile(RenameFile renameFile) {
+        // "documentChanges": [
+        //  {
+        //    "oldUri": "file://.../foo.clj",
+        //    "newUri": "file://.../bar.clj",
+        //    "kind": "rename"
+        //  }
+        //]
+        try {
+            VirtualFile resource = findResourceFor(renameFile.getOldUri());
+            if (resource != null) {
+                // The IJ rename works only with a file name which is hosted in the same directory
+                // as the old file
+                var path = Paths.get(URI.create(renameFile.getNewUri()));
+                String newFileName = path.toFile().getName();
```

Comment:
What if new filename already exists?

Reply:
It should fail, but could we see this usecase in ananother PR since I have not this usecase with all LS that I hav etried (scala, rust, go, clojure,typescript)